common/dbus.hpp

```diff
@@ -0,0 +1,103 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#ifndef MENDER_COMMON_DBUS_HPP
+#define MENDER_COMMON_DBUS_HPP
+
+#include <config.h>
+
+#include <functional>
+#include <memory>
+#include <string>
+
+#include <common/error.hpp>
+#include <common/expected.hpp>
+#include <common/events.hpp>
+
+#ifdef MENDER_USE_ASIO_LIBDBUS
+#include <dbus/dbus.h>
+#endif // MENDER_USE_ASIO_LIBDBUS
+
+namespace mender {
+namespace common {
+namespace dbus {
+
+namespace error = mender::common::error;
+namespace expected = mender::common::expected;
+namespace events = mender::common::events;
+
+using namespace std;
+
+enum DBusErrorCode {
+	NoError = 0,
+	ConnectionError,
+	MessageError,
+	ReplyError,
+	ValueError,
+};
+
+class DBusErrorCategoryClass : public std::error_category {
+public:
+	const char *name() const noexcept override;
+	string message(int code) const override;
+};
+extern const DBusErrorCategoryClass DBusErrorCategory;
+
+error::Error MakeError(DBusErrorCode code, const string &msg);
+
+// TODO: template <typename reply_type> or
+//       DBusCallReplyHandler<DBusString> or
+//       DBusCallStringReplyHandler?
+using DBusCallReplyHandler = function<void(expected::ExpectedString)>;
+
+class DBusClient : events::EventLoopObject {
+public:
+	DBusClient(events::EventLoop &loop) :
```

Comment:
Nit: explicit?

Reply:
I wish there was a `-fno-fucking-implicit-anything` compiler option. But yes, we shouldn't (I guess?!) allow things like `DBusClient client = loop` or using `loop` in places where an actual `DBusClient` instance is expected.